---
title: "Compile molecular subtyping results"
output: 
  html_notebook:
    toc: true
    toc_float: true
author: Jaclyn Taroni for CCDL, Jo Lynne Rokita for D3b, Zhuangzhuang Geng for D3b
date: 2020
params:
   is_ci: FALSE
---

The purpose of this notebook is to aggregate molecular subtyping results from the following mature analysis modules:

* [`molecular-subtyping-EWS`](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/tree/master/analyses/molecular-subtyping-EWS)
* [`molecular-subtyping-HGG`](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/tree/master/analyses/molecular-subtyping-HGG)
* [`molecular-subtyping-LGAT`](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/tree/master/analyses/molecular-subtyping-LGAT)
* [`molecular-subtyping-embryonal`](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/tree/master/analyses/molecular-subtyping-embryonal)
* [`molecular-subtyping-CRANIO`](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/tree/master/analyses/molecular-subtyping-CRANIO)
* [`molecular-subtyping-EPN`](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/tree/master/analyses/molecular-subtyping-EPN)
* [`molecular-subtyping-MB`](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/tree/master/analyses/molecular-subtyping-MB)
* [`molecular-subtyping-neurocytoma`](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/tree/master/analyses/molecular-subtyping-neurocytoma)
* [`molecular-subtyping-chordoma`](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/tree/master/analyses/molecular-subtyping-chordoma)
* [`molecular-subtyping-ATRT`](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/tree/dev/analyses/molecular-subtyping-ATRT)
* [`molecular-subtyping-NBL`](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/tree/dev/analyses/molecular-subtyping-NBL)


## Set up

```{r}
# This logic is the same as analyses/gene-set-enrichment-analysis/02-model-gsea.Rmd
# Assigning params$is_ci to running_in_ci avoids a locked binding error
running_in_ci <- params$is_ci

# Are we testing? In case of a non 0/1 number, we recast as logical, and then 
# ensure logical.
if (running_in_ci %in% c(0,1)) running_in_ci <- as.logical(running_in_ci)
if (!(is.logical(running_in_ci)))
{
  stop("\n\nERROR: The parameter `is_ci` should be FALSE/TRUE (or 0/1).")
}
```

### Libraries and functions

```{r}
library(tidyverse)
```

Create function to separate and rebind bs_ids from results files to mimic histologies file
```{r}
bind_exp_strategies <- function(results_df) {
  # Many of the molecular subtyping tables contain a 
  # Kids_First_Biospecimen_ID_RNA and Kids_First_Biospecimen_ID_DNA column,
  # this function separates those out and binds the rows to mirror how 
  # histologies.tsv is organized
  #
  # Args:
  #   results_df: a molecular subtyping table with Kids_First_Biospecimen_ID_RNA,
  #               and Kids_First_Biospecimen_ID_DNA columns
  # Returns:
  #   A data.frame where each biospecimen is a row
  
  # So far, files have one of two sets of columns - this might change in the future or we might need to think of how to better handle
  dna_rna_cols <- c("Kids_First_Biospecimen_ID_DNA", "Kids_First_Biospecimen_ID_RNA")
  long_cols <- c("Kids_First_Biospecimen_ID_methyl", "Kids_First_Biospecimen_ID_DNA_wgs, Kids_First_Biospecimen_ID_DNA_panel, Kids_First_Biospecimen_ID_RNA_panel", "Kids_First_Biospecimen_ID_RNA")
  
    # if only DNA and RNA present:
  if (dna_rna_cols %in% names(results_df)) {
    rna_df <- results_df %>%
    dplyr::select(-Kids_First_Biospecimen_ID_DNA) %>%
    dplyr::filter(!is.na(Kids_First_Biospecimen_ID_RNA)) %>%
    dplyr::rename(Kids_First_Biospecimen_ID = Kids_First_Biospecimen_ID_RNA)
  
    dna_df <- results_df %>%
    dplyr::select(-Kids_First_Biospecimen_ID_RNA) %>%
    dplyr::filter(!is.na(Kids_First_Biospecimen_ID_DNA)) %>%
    dplyr::rename(Kids_First_Biospecimen_ID = Kids_First_Biospecimen_ID_DNA)
    
    return(bind_rows(rna_df, dna_df) %>% distinct())
  }
  
    # if all the things - DNA, RNA, methyl, DNA and RNA panel - are present:
  if (long_cols %in% names(results_df)) {
    rna_df <- results_df %>%
    dplyr::select(-c(Kids_First_Biospecimen_ID_methyl, Kids_First_Biospecimen_ID_DNA_panel, Kids_First_Biospecimen_ID_RNA_panel, Kids_First_Biospecimen_ID_DNA_wgs)) %>%
    dplyr::filter(!is.na(Kids_First_Biospecimen_ID_RNA)) %>%
    dplyr::rename(Kids_First_Biospecimen_ID = Kids_First_Biospecimen_ID_RNA)
    
    dna_df <- results_df %>%
    dplyr::select(-c(Kids_First_Biospecimen_ID_methyl, Kids_First_Biospecimen_ID_DNA_panel, Kids_First_Biospecimen_ID_RNA_panel, Kids_First_Biospecimen_ID_RNA)) %>%
    dplyr::filter(!is.na(Kids_First_Biospecimen_ID_DNA_wgs)) %>%
    dplyr::rename(Kids_First_Biospecimen_ID = Kids_First_Biospecimen_ID_DNA_wgs)
    
    rna_panel_df <- results_df %>%
    dplyr::select(-c(Kids_First_Biospecimen_ID_methyl, Kids_First_Biospecimen_ID_DNA_wgs, Kids_First_Biospecimen_ID_DNA_panel, Kids_First_Biospecimen_ID_RNA)) %>%
    dplyr::filter(!is.na(Kids_First_Biospecimen_ID_RNA_panel)) %>%
    dplyr::rename(Kids_First_Biospecimen_ID = Kids_First_Biospecimen_ID_RNA_panel)
    
    dna_panel_df <- results_df %>%
    dplyr::select(-c(Kids_First_Biospecimen_ID_methyl, Kids_First_Biospecimen_ID_DNA_wgs, Kids_First_Biospecimen_ID_RNA_panel, Kids_First_Biospecimen_ID_RNA)) %>%
    dplyr::filter(!is.na(Kids_First_Biospecimen_ID_DNA_panel)) %>%
    dplyr::rename(Kids_First_Biospecimen_ID = Kids_First_Biospecimen_ID_DNA_panel)
    
    methyl_df <- results_df %>%
    dplyr::select(-c(Kids_First_Biospecimen_ID_DNA_panel, Kids_First_Biospecimen_ID_DNA_wgs, Kids_First_Biospecimen_ID_RNA_panel, Kids_First_Biospecimen_ID_RNA)) %>%
    dplyr::filter(!is.na(Kids_First_Biospecimen_ID_methyl)) %>%
    dplyr::rename(Kids_First_Biospecimen_ID = Kids_First_Biospecimen_ID_methyl)
    
    return(bind_rows(rna_df, dna_df, rna_panel_df, dna_panel_df, methyl_df) %>% distinct())
  }
    
  # sometimes there will be multiple RNA-seq assays per single DNA-seq assay
  # or vice versa, which is why this distinct is necessary
  
}
```

### Directories and files

```{r}
data_dir <- file.path("..", "..", "data")
analyses_dir <- ".."

# directories for upstream subtyping modules
cranio_dir <- file.path(analyses_dir, "molecular-subtyping-CRANIO")
ews_dir <- file.path(analyses_dir, "molecular-subtyping-EWS")
epn_dir <- file.path(analyses_dir, "molecular-subtyping-EPN")
hgg_dir <- file.path(analyses_dir, "molecular-subtyping-HGG")
lgat_dir <- file.path(analyses_dir, "molecular-subtyping-LGAT")
mb_dir <- file.path(analyses_dir, "molecular-subtyping-MB")
embryonal_dir <- file.path(analyses_dir, "molecular-subtyping-embryonal")
neurocytoma_dir <- file.path(analyses_dir, "molecular-subtyping-neurocytoma")
chordoma_dir <- file.path(analyses_dir, "molecular-subtyping-chordoma")
atrt_dir <- file.path(analyses_dir, "molecular-subtyping-ATRT")
nbl_dir <- file.path(analyses_dir, "molecular-subtyping-NBL")

# the folder that contains the tabular results is standardized across modules
results_dir <- "results"
# create the results directory for this module if it does not yet exist
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}
```

#### Input files

When we run this locally, we want to tie it to a specific version of the histologies file _prior_ to feedback we received from pathology that is recorded on [#609](https://github.com/AlexsLemonade/OpenPBTA-analysis/issues/609).

```{r}
if (running_in_ci) {
  histologies_file <- file.path(data_dir, "histologies-base.tsv")
} else {
  histologies_file <- file.path(data_dir, 
                                "histologies-base.tsv")
}

join_cols <- c("Kids_First_Biospecimen_ID",
               "short_histology",
               "broad_histology",
               "Notes")

```

Results files from individual modules.

```{r}
cranio_results_file <- file.path(cranio_dir, results_dir, "CRANIO_molecular_subtype.tsv")
ews_results_file <- file.path(ews_dir, results_dir, "EWS_results.tsv")
epn_results_file <- file.path(epn_dir, results_dir, "EPN_all_data_withsubgroup.tsv")
hgg_results_file <- file.path(hgg_dir, results_dir, "HGG_molecular_subtype.tsv")
lgat_results_file <- file.path(lgat_dir, results_dir, "lgat_subtyping.tsv")
mb_results_file <- file.path(mb_dir, results_dir, "MB_molecular_subtype.tsv")
embryonal_results_file <- file.path(embryonal_dir, results_dir,
                                    "embryonal_tumor_molecular_subtypes.tsv")
neurocytoma_results_file <- file.path(neurocytoma_dir, results_dir, "neurocytoma_subtyping.tsv")
# reading in smarcb1 status only, subtyping will come through pathology review in step 3
chordoma_results_file <- file.path(chordoma_dir, results_dir, "chordoma_smarcb1_status.tsv")
atrt_results_file <- file.path(atrt_dir, results_dir, "ATRT-molecular-subtypes.tsv")
nbl_results_file <- file.path(nbl_dir, results_dir, "neuroblastoma_molecular_subtypes.tsv")
```

#### Output file

```{r}
output_file <- file.path(results_dir, "compiled_molecular_subtypes.tsv")
```

## Read in data

```{r message=FALSE}
# split 
histologies_df <- read_tsv(histologies_file, guess_max = 100000, col_types = cols(.default = "c"))
cranio_results_df <- read_tsv(cranio_results_file)
ews_results_df <- read_tsv(ews_results_file)
epn_results_df <- read_tsv(epn_results_file)
hgg_results_df <- read_tsv(hgg_results_file)
lgat_results_df <- read_tsv(lgat_results_file)
mb_results_df <- read_tsv(mb_results_file)
neurocytoma_results_df <- read_tsv(neurocytoma_results_file)
embryonal_results_df <- read_tsv(embryonal_results_file)
chordoma_results_df <- read_tsv(chordoma_results_file)
atrt_results_df <- read_tsv(atrt_results_file)
nbl_results_df <- read_tsv(nbl_results_file)
```

## Compile the subtyping resutls

### Handling non-ATRT/non-MB embryonal tumors

```{r non atrt/non mb}
embryonal_results_df <- bind_exp_strategies(embryonal_results_df) %>%
  inner_join(select(histologies_df,join_cols),
             by = "Kids_First_Biospecimen_ID") %>%
  mutate(integrated_diagnosis = case_when(molecular_subtype == "CNS Embryonal, NOS"~ "CNS Embryonal tumor, NEC/NOS",
                                          molecular_subtype == "CNS HGNET-MN1"~ "CNS Embryonal tumor, HGNET-MN1",
                                          molecular_subtype == "CNS NB-FOXR2" ~ "CNS neuroblastoma",
                                          molecular_subtype == "ETMR, C19MC-altered"~ "Embryonal tumor with multilayer rosettes, C19MC-altered",
                                          molecular_subtype == "ETMR, NOS"~"Embryonal tumor with multilayer rosettes, NOS",
                                          TRUE ~ NA_character_), 
        short_histology = 
          if_else(molecular_subtype %in% c("ETMR, C19MC-altered", "ETMR, NOS"),
                "ETMR", "Embryonal tumor"),
        broad_histology = "Embryonal tumor",
        Notes = if_else(!is.na(integrated_diagnosis), "Updated via OpenPedCan subtyping", Notes))

```

### Handling EWS

```{r}
# Add EWS integrated diagnosis, broad histology, short histology
ews_results_df <- bind_exp_strategies(ews_results_df) %>%
  inner_join(select(histologies_df,join_cols),
             by = "Kids_First_Biospecimen_ID") %>%
  mutate(integrated_diagnosis = "Ewing sarcoma",
         broad_histology = "Mesenchymal non-meningothelial tumor",
         short_histology = "EWS",
         Notes = if_else(!is.na(integrated_diagnosis), "Updated via OpenPedCan subtyping", Notes))
```

### Handling HGG

```{r}
hgg_results_df <- bind_exp_strategies(hgg_results_df) %>%
  inner_join(select(histologies_df,join_cols),
             by = "Kids_First_Biospecimen_ID") %>%
  mutate(integrated_diagnosis = case_when(grepl("DMG, H3 K28", molecular_subtype) ~ "Diffuse midline glioma, H3 K28-mutant",
                                          grepl("HGG, H3 G35", molecular_subtype) ~ "Diffuse hemispheric glioma, H3 G35-mutant",
                                          grepl("HGG, H3 wildtype", molecular_subtype) ~ "High-grade glioma, IDH-wildtype and H3-wildtype",
                                          grepl("HGG, IDH", molecular_subtype) ~ "High-grade glioma, IDH-mutant",
                                          # account for IGH subtypes
                                          grepl("IHG, NTRK-altered", molecular_subtype) ~ "Infant-type hemispheric glioma, NTRK-altered",
                                          grepl("IHG, ALK-altered", molecular_subtype) ~ "Infant-type hemispheric glioma, ALK-altered",
                                          grepl("IHG, ROS1-altered", molecular_subtype) ~ "Infant-type hemispheric glioma, ROS1-altered",
                                          grepl("IHG, MET-altered", molecular_subtype) ~ "Infant-type hemispheric glioma, MET-altered",
                                          TRUE~ NA_character_),
                  broad_histology = "Diffuse astrocytic and oligodendroglial tumor",
                  short_histology = "HGAT",
         Notes = if_else(!is.na(integrated_diagnosis), "Updated via OpenPedCan subtyping", Notes))
```

### Handling LGAT
Within LGAT, we will add integrated_diagnosis,broad_histology and short_histology for Ganglioglioma (GNG) and Subependymal Giant Cell Astrocytoma (SEGA) samples.
For Glial neuronal tumors (GNT) will be updated these values using pathology_free_text_diagnosis in the next step.

```{r}
lgat_results_df <- bind_exp_strategies(lgat_results_df) %>%
  inner_join(select(histologies_df,c(join_cols,
                                     pathology_diagnosis)),
             by = "Kids_First_Biospecimen_ID") %>%
  ## Integrated diagnosis update
  mutate(integrated_diagnosis = case_when(
    str_detect(molecular_subtype, "To be classified") ~ NA_character_,
    grepl("^GNG, ",molecular_subtype) ~ str_replace(molecular_subtype, "GNG, ", "Ganglioglioma, "),
    grepl("^SEGA, ",molecular_subtype) ~ str_replace(molecular_subtype, "SEGA, ", "Subependymal Giant Cell Astrocytoma, "),
    TRUE ~ str_replace(molecular_subtype, "LGG, ", "Low-grade glioma, ")
  ),
  ## Broad histology update
  broad_histology = "Low-grade astrocytic tumor",
  ## Short histology update
  short_histology = case_when(
    grepl("^GNG, ",molecular_subtype) ~ "Ganglioglioma",
    TRUE ~ "LGAT"),
  Notes = if_else(!is.na(integrated_diagnosis), "Updated via OpenPedCan subtyping", Notes)
  )

```

### Handling EPN 

```{r}
epn_results_df <- bind_exp_strategies(epn_results_df) %>%
  inner_join(select(histologies_df,join_cols),
           by = "Kids_First_Biospecimen_ID") %>%
  mutate(molecular_subtype = subgroup) %>%
  mutate(molecular_subtype = if_else(is.na(molecular_subtype), "EPN, To be classified", molecular_subtype), 
         integrated_diagnosis = case_when(molecular_subtype == "EPN, PF A" ~ "Posterior fossa group A (PFA) ependymoma",
                                          molecular_subtype == "EPN, ST ZFTA" ~ "Supratentorial ependymoma, ZFTA fusion-positive",
                                          molecular_subtype == "EPN, ST YAP1" ~ "Supratentorial ependymoma, YAP1 fusion-positive", 
                                          # add new EPN subtypes
                                          molecular_subtype == "EPN, SP" ~ "Spinal ependymoma, NOS",
                                          molecular_subtype == "EPN, SP-MYCN" ~ "Spinal ependymoma, MYCN-amplified",
                                          TRUE~ NA_character_),
         broad_histology = "Ependymal tumor",
         short_histology = "Ependymoma",
         Notes = if_else(!is.na(integrated_diagnosis), "Updated via OpenPedCan subtyping", Notes))
```

### Handling MB

```{r}
mb_results_df <- bind_exp_strategies(mb_results_df) %>%
  inner_join(select(histologies_df,join_cols),
            by = "Kids_First_Biospecimen_ID") %>%
  mutate(integrated_diagnosis = case_when(molecular_subtype == "MB, SHH" ~ "Medulloblastoma, SHH-activated",
                                          molecular_subtype == "MB, WNT" ~ "Medulloblastoma, WNT-activated",
                                          molecular_subtype == "MB, Group3" ~ "Medulloblastoma, group 3",
                                          molecular_subtype == "MB, Group4" ~ "Medulloblastoma, group 4", 
                                          TRUE ~NA_character_),
         broad_histology = "Embryonal tumor",
         short_histology = "Medulloblastoma",
         Notes = if_else(!is.na(integrated_diagnosis), "Subtype based on prediction;Updated via OpenPedCan subtyping", Notes))
```


### Handling CRANIO

```{r}
cranio_results_df <- bind_exp_strategies(cranio_results_df) %>%
  inner_join(select(histologies_df,join_cols),
             by = "Kids_First_Biospecimen_ID") %>%
  mutate(integrated_diagnosis = case_when(molecular_subtype == "CRANIO, ADAM" ~"Adamantinomatous craniopharyngioma",
                                          molecular_subtype == "CRANIO, PAP" ~"Papillary craniopharyngioma", 
                                          TRUE ~ NA_character_),
         broad_histology = "Tumors of sellar region",
         short_histology = "Craniopharyngioma",
         Notes = if_else(!is.na(integrated_diagnosis), "Updated via OpenPedCan subtyping", Notes))
```

### Handling Neurocytoma
```{r}
neurocytoma_results_df <- neurocytoma_results_df %>%
  inner_join(select(histologies_df,join_cols),
            by = "Kids_First_Biospecimen_ID") %>%
  mutate(harmonized_diagnosis = case_when(molecular_subtype == "CNC"~ "Central Neurocytoma",
                                          molecular_subtype == "EVN" ~"Extraventricular Neurocytoma", 
                                        TRUE ~ NA_character_),
         broad_histology = "Neuronal and mixed neuronal-glial tumor",
         short_histology = "Neurocytoma",
         Notes = if_else(!is.na(harmonized_diagnosis), "Updated via OpenPedCan subtyping", Notes))
```

### Handling Chordoma
```{r}

chordoma_results_df <- bind_exp_strategies(chordoma_results_df) %>%
  inner_join(select(histologies_df,join_cols),
             by = "Kids_First_Biospecimen_ID") %>%
  # will update subtype and integrated diagnosis in step 3
  mutate(molecular_subtype = NA_character_,
         integrated_diagnosis = NA_character_,
         broad_histology = "Chordoma",
         short_histology = "Chordoma",
         Notes = if_else(!is.na(integrated_diagnosis), "Updated via OpenPedCan subtyping", Notes))

```

### Handling ATRT
```{r}
atrt_results_df <-  bind_exp_strategies(atrt_results_df) %>%
  inner_join(select(histologies_df,join_cols),
             by = "Kids_First_Biospecimen_ID") %>%
  mutate(integrated_diagnosis = case_when(molecular_subtype == "ATRT, MYC" ~ "Atypical Teratoid Rhabdoid Tumor, MYC",
                                          molecular_subtype == "ATRT, TYR" ~ "Atypical Teratoid Rhabdoid Tumor, TYR", 
                                          molecular_subtype == "ATRT, SHH" ~ "Atypical Teratoid Rhabdoid Tumor, SHH", 
                                          TRUE ~ NA_character_),
         broad_histology = "Embryonal tumor",
         short_histology = "ATRT",
         Notes = if_else(!is.na(integrated_diagnosis), "Subtype based on methylation;Updated via OpenPedCan subtyping", Notes))

```
### leave this section  for NBL module


```{r}
nbl_results_df <- bind_exp_strategies(nbl_results_df) %>%
  inner_join(select(histologies_df, c("Kids_First_Participant_ID", "sample_id", join_cols)),
             by = "Kids_First_Biospecimen_ID") %>%
  mutate(integrated_diagnosis = case_when(molecular_subtype == "NBL, MYCN amplified" ~ paste0(short_histology, ", MYCN-amplified"),
                                          molecular_subtype == "NBL, MYCN non-amplified" ~ paste0(short_histology, ", MYCN-non-amplified"), 
                                          TRUE ~ NA_character_),
         broad_histology = "Embryonal tumor",
         Notes = if_else(!is.na(integrated_diagnosis), "Updated via OpenPedCan subtyping", Notes))
```
### All results

Compile results, sort, and write to file

```{r}
all_results_df <- bind_rows(embryonal_results_df,
                            ews_results_df,
                            hgg_results_df,
                            lgat_results_df,
                            epn_results_df,
                            cranio_results_df,
                            mb_results_df,
                            neurocytoma_results_df,
                            chordoma_results_df,
                            atrt_results_df, 
                            nbl_results_df, 
                            ) %>%
  select(Kids_First_Participant_ID, sample_id, Kids_First_Biospecimen_ID, molecular_subtype,
         integrated_diagnosis, short_histology, broad_histology, Notes) %>%
  # Cleanup a few Notes which have changed since last time
  # Remove this because those were taken out of EWS
  mutate(Notes = case_when(Notes == "Reclassified due to presence of hallmark EWS fusions"~ NA_character_,
                          Notes == "Subtype based on prediction"~ NA_character_,
                          TRUE ~ Notes)) %>%
  arrange(Kids_First_Participant_ID, sample_id) %>%
  write_tsv(output_file)
```

## Session Info

```{r}
sessionInfo()
```